File: back/world/recording.go
package world

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"os"
	"sort"
	"strconv"
)

type Recording struct {
	Name            string
	World           *World
	SecondsPerFrame float32
	EndSeconds      int
}

type RecordingFile struct {
	SecondsPerFrame float32
	TotalFrames     uint32
	Frames          map[string]Frame
}

func NewRecording(name string, world *World, secondsPerFrame float32, endSeconds int) *Recording {
	return &Recording{Name: name, World: world, SecondsPerFrame: secondsPerFrame, EndSeconds: endSeconds}
}

func (r *Recording) Simulate() {
	fmt.Println("Simulating recording...")

	frames := make(map[string]Frame)
	frameInterval := r.SecondsPerFrame
	totalFrames := int(float32(r.EndSeconds)/r.SecondsPerFrame) + 1

	fmt.Println("Total frames:", totalFrames)

	chunkIndex := 0
	exePath, _ := os.Getwd()
	logDir := fmt.Sprintf("%s/logs/%s", exePath, r.Name)
	os.MkdirAll(logDir, 0755)
	//clear before starting
	//os.RemoveAll(logDir)

	for i := 0; i < totalFrames; i++ {
		fmt.Println("Frame:", i)

		target := float32(i) * frameInterval
		for r.World.CurTime < target {
			r.World.Update()
		}
		frame := r.World.Flatten()
		frames[fmt.Sprintf("%.3f", target)] = frame

		if len(frames) == 1000 || i == totalFrames-1 {
			fileStruct := RecordingFile{
				SecondsPerFrame: r.SecondsPerFrame,
				TotalFrames:     uint32(totalFrames),
				Frames:          frames,
			}

			logPath := fmt.Sprintf("%s/%d.bin", logDir, chunkIndex)
			file, _ := os.Create(logPath)
			data := EncodeRecordingBinary(fileStruct)
			file.Write(data)
			file.Close()

			frames = make(map[string]Frame)
			chunkIndex++
		}
	}
}

func EncodeRecordingBinary(r RecordingFile) []byte {
	buf := &bytes.Buffer{}

	buf.Write([]byte{'S', 'I', 'M', 'B'})
	binary.Write(buf, binary.LittleEndian, r.SecondsPerFrame)
	binary.Write(buf, binary.LittleEndian, r.TotalFrames)

	keys := make([]float64, 0, len(r.Frames))
	for k := range r.Frames {
		f, _ := strconv.ParseFloat(k, 32)
		keys = append(keys, f)
	}
	sort.Float64s(keys)

	binary.Write(buf, binary.LittleEndian, uint32(len(keys)))
	for _, key := range keys {
		t := float32(key)
		binary.Write(buf, binary.LittleEndian, t)

		frame := r.Frames[fmt.Sprintf("%.3f", t)]
		count := uint32(len(frame.IDArray))
		binary.Write(buf, binary.LittleEndian, count)

		for i := 0; i < int(count); i++ {
			id := frame.IDArray[i]
			binary.Write(buf, binary.LittleEndian, uint32(len(id)))
			buf.WriteString(id)

			pos := frame.PositionArrays[i]
			for j := 0; j < 3; j++ {
				binary.Write(buf, binary.LittleEndian, pos[j])
			}

			verts := frame.VertexArrays[i]
			binary.Write(buf, binary.LittleEndian, uint32(len(verts)))
			for _, v := range verts {
				binary.Write(buf, binary.LittleEndian, v)
			}

			inds := frame.IndexArrays[i]
			binary.Write(buf, binary.LittleEndian, uint32(len(inds)))
			for _, ind := range inds {
				binary.Write(buf, binary.LittleEndian, ind)
			}

			color := frame.ColorArray[i]
			binary.Write(buf, binary.LittleEndian, uint32(len(color)))
			buf.WriteString(color)
		}
	}

	return buf.Bytes()
}

func DecodeRecordingBinary(data []byte) RecordingFile {
	buf := bytes.NewReader(data)
	magic := make([]byte, 4)
	buf.Read(magic)

	var spf float32
	binary.Read(buf, binary.LittleEndian, &spf)

	var total uint32
	binary.Read(buf, binary.LittleEndian, &total)

	var frameCount uint32
	binary.Read(buf, binary.LittleEndian, &frameCount)

	frames := make(map[string]Frame)
	for i := 0; i < int(frameCount); i++ {
		var t float32
		binary.Read(buf, binary.LittleEndian, &t)
		var objCount uint32
		binary.Read(buf, binary.LittleEndian, &objCount)

		idArray := make([]string, objCount)
		posArray := make([][]float32, objCount)
		vertArray := make([][]float32, objCount)
		indArray := make([][]uint32, objCount)
		colorArray := make([]string, objCount)

		for j := 0; j < int(objCount); j++ {
			var l uint32
			binary.Read(buf, binary.LittleEndian, &l)
			idBytes := make([]byte, l)
			buf.Read(idBytes)
			idArray[j] = string(idBytes)

			p := make([]float32, 3)
			for k := 0; k < 3; k++ {
				binary.Read(buf, binary.LittleEndian, &p[k])
			}
			posArray[j] = p

			binary.Read(buf, binary.LittleEndian, &l)
			verts := make([]float32, l)
			for k := 0; k < int(l); k++ {
				binary.Read(buf, binary.LittleEndian, &verts[k])
			}
			vertArray[j] = verts

			binary.Read(buf, binary.LittleEndian, &l)
			inds := make([]uint32, l)
			for k := 0; k < int(l); k++ {
				binary.Read(buf, binary.LittleEndian, &inds[k])
			}
			indArray[j] = inds

			binary.Read(buf, binary.LittleEndian, &l)
			colorBytes := make([]byte, l)
			buf.Read(colorBytes)
			colorArray[j] = string(colorBytes)
		}

		frames[fmt.Sprintf("%.3f", t)] = Frame{
			CurTime:        t,
			IDArray:        idArray,
			PositionArrays: posArray,
			VertexArrays:   vertArray,
			IndexArrays:    indArray,
			ColorArray:     colorArray,
		}
	}

	return RecordingFile{SecondsPerFrame: spf, TotalFrames: total, Frames: frames}
}

File: back/recording.py
"""
Recording system - matches Go binary format exactly.
"""
import struct
from pathlib import Path
from world import World


class Recording:
    def __init__(self, name: str, world: World, seconds_per_frame: float, end_seconds: int):
        """
        Args:
            name: Recording name
            world: World to simulate
            seconds_per_frame: Time between recorded frames
            end_seconds: Total simulation duration
        """
        self.name = name
        self.world = world
        self.seconds_per_frame = seconds_per_frame
        self.end_seconds = end_seconds

        # Create recording directory (one level up from python-back)
        self.recording_dir = Path(__file__).parent.parent / 'logs' / name
        self.recording_dir.mkdir(parents=True, exist_ok=True)

    def simulate(self):
        """Run simulation and save recordings."""
        print("Simulating recording...")

        total_frames = int(self.end_seconds / self.seconds_per_frame) + 1
        print(f"Total frames: {total_frames}")

        frames = {}  # Dict of time -> frame_data
        chunk_index = 0

        for i in range(total_frames):
            if i % 100 == 0:
                print(f"Frame: {i}")

            # Run simulation until target time
            target_time = i * self.seconds_per_frame
            while self.world.time < target_time:
                self.world.update()

            # Record frame
            frame_data = self.world.get_frame_data()
            time_key = f"{target_time:.3f}"
            frames[time_key] = frame_data

            # Save chunk every 1000 frames or at end
            if len(frames) == 1000 or i == total_frames - 1:
                self._save_chunk(chunk_index, frames, total_frames)
                frames = {}
                chunk_index += 1

        print(f"Recording complete! Saved {chunk_index} chunks")

    def _save_chunk(self, chunk_index: int, frames: dict, total_frames: int):
        """Save chunk in Go-compatible binary format."""
        chunk_path = self.recording_dir / f"{chunk_index}.bin"

        with open(chunk_path, 'wb') as f:
            # Magic header
            f.write(b'SIMB')

            # Seconds per frame
            f.write(struct.pack('<f', self.seconds_per_frame))

            # Total frames in entire recording
            f.write(struct.pack('<I', total_frames))

            # Number of frames in this chunk
            f.write(struct.pack('<I', len(frames)))

            # Sort frames by time
            sorted_times = sorted([float(k) for k in frames.keys()])

            for time in sorted_times:
                time_key = f"{time:.3f}"
                frame = frames[time_key]

                # Write frame time
                f.write(struct.pack('<f', time))

                # Number of objects
                num_objects = len(frame['id_array'])
                f.write(struct.pack('<I', num_objects))

                # Write each object
                for i in range(num_objects):
                    # ID (length-prefixed string)
                    obj_id = frame['id_array'][i].encode('utf-8')
                    f.write(struct.pack('<I', len(obj_id)))
                    f.write(obj_id)

                    # Position (3 floats)
                    pos = frame['position_arrays'][i]
                    for j in range(3):
                        f.write(struct.pack('<f', pos[j]))

                    # Rotation matrix (9 floats, row-major 3x3)
                    rot = frame['rotation_arrays'][i]
                    for j in range(9):
                        f.write(struct.pack('<f', rot[j]))

                    # Vertices (count + floats)
                    verts = frame['vertex_arrays'][i]
                    f.write(struct.pack('<I', len(verts)))
                    for v in verts:
                        f.write(struct.pack('<f', v))

                    # Indices (count + uints)
                    inds = frame['index_arrays'][i]
                    f.write(struct.pack('<I', len(inds)))
                    for idx in inds:
                        f.write(struct.pack('<I', idx))

                    # Color (length-prefixed string)
                    color = frame['color_array'][i].encode('utf-8')
                    f.write(struct.pack('<I', len(color)))
                    f.write(color)

        print(f"Saved chunk {chunk_index} with {len(frames)} frames")